### PR TITLE
[GISel] Add support for scalable vectors in getLCMType

### PR DESCRIPTION
**Pull request by michaelmaitland**
This function can be called from buildCopyToRegs where at least one of the types is a scalable vector type. This function crashed because it did not know how to handle scalable vector types.

This patch extends the functionality of getLCMType to handle when at least one of the types is a scalable vector. getLCMType between a fixed and scalable vector is not implemented since the docstring of the function explains that getLCMType is used to build MERGE/UNMERGE instructions and we will never build a MERGE/UNMERGE between fixed and scalable vectors.

### CODE REVIEW

**Comment by tschuett** [comment]
G_UNMERGE_VALUES, G_MERGE_VALUES, and G_BUILD_VECTOR only work with fixed length vectors. Is the root cause of your crash somewhere else?

**Comment by michaelmaitland** [comment]
> G_UNMERGE_VALUES, G_MERGE_VALUES, and G_BUILD_VECTOR only work with fixed length vectors. Is the root cause of your crash somewhere else?

I am not so sure that I agree with `G_UNMERGE_VALUES, G_MERGE_VALUES only work with fixed length vectors`. Check this test out:
```
; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
; RUN: llc -mtriple=riscv32 -mattr=+v -global-isel -stop-after=irtranslator \
; RUN:   -verify-machineinstrs < %s | FileCheck -check-prefixes=RV32 %s
; RUN: llc -mtriple=riscv64 -mattr=+v -global-isel -stop-after=irtranslator \
; RUN:   -verify-machineinstrs < %s | FileCheck -check-prefixes=RV64 %s
define <vscale x 16 x i64> @shufflevector_nxv16i64_1() {
  ; RV32-LABEL: name: shufflevector_nxv16i64_1
  ; RV32: bb.1 (%ir-block.0):
  ; RV32-NEXT:   [[DEF:%[0-9]+]]:_(<vscale x 16 x s64>) = G_IMPLICIT_DEF
  ; RV32-NEXT:   [[SHUF:%[0-9]+]]:_(<vscale x 16 x s64>) = G_SHUFFLE_VECTOR [[DEF]](<vscale x 16 x s64>), [[DEF]], hufflemask(undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef)
  ; RV32-NEXT:   [[UV:%[0-9]+]]:_(<vscale x 8 x s64>), [[UV1:%[0-9]+]]:_(<vscale x 8 x s64>) = G_UNMERGE_VALUES [SHUF]](<vscale x 16 x s64>)
  ; RV32-NEXT:   $v8m8 = COPY [[UV]](<vscale x 8 x s64>)
  ; RV32-NEXT:   $v16m8 = COPY [[UV1]](<vscale x 8 x s64>)
  ; RV32-NEXT:   PseudoRET implicit $v8m8, implicit $v16m8
  ;
  ; RV64-LABEL: name: shufflevector_nxv16i64_1
  ; RV64: bb.1 (%ir-block.0):
  ; RV64-NEXT:   [[DEF:%[0-9]+]]:_(<vscale x 16 x s64>) = G_IMPLICIT_DEF
  ; RV64-NEXT:   [[SHUF:%[0-9]+]]:_(<vscale x 16 x s64>) = G_SHUFFLE_VECTOR [[DEF]](<vscale x 16 x s64>), [[DEF]], shufflemask(undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef, undef)
  ; RV64-NEXT:   [[UV:%[0-9]+]]:_(<vscale x 8 x s64>), [[UV1:%[0-9]+]]:_(<vscale x 8 x s64>) = G_UNMERGE_VALUES [[SHUF]](<vscale x 16 x s64>)
  ; RV64-NEXT:   $v8m8 = COPY [[UV]](<vscale x 8 x s64>)
  ; RV64-NEXT:   $v16m8 = COPY [[UV1]](<vscale x 8 x s64>)
  ; RV64-NEXT:   PseudoRET implicit $v8m8, implicit $v16m8
  %a = shufflevector <vscale x 16 x i64> undef, <vscale x 16 x i64> undef, <vscale x 16 x i32> undef
  ret <vscale x 16 x i64> %a
}
```

There is definitely G_UNMERGE_VALUES getting generated. This test was generated with my fixing of get LCM/GCD types updated.

**Comment by tschuett** [comment]
This looks odd. In your examples, you try to unmerge a scalable vector. In how many register do want to unmerge  the scalable vector. From the documentation:

```
%bits_0_7:(s8), %bits_8_15:(s8),
    %bits_16_23:(s8), %bits_24_31:(s8) = G_UNMERGE_VALUES %0:(s32)
```
We unmerge 32 bits into 4 registers of 8 bits. For scalable  vectors, the number of output registers is unknown.

**Comment by michaelmaitland** [comment]
> This looks odd. In your examples, you try to unmerge a scalable vector. In how many register do want to unmerge the scalable vector. From the documentation:
> 
> ```
> %bits_0_7:(s8), %bits_8_15:(s8),
>     %bits_16_23:(s8), %bits_24_31:(s8) = G_UNMERGE_VALUES %0:(s32)
> ```
> 
> We unmerge 32 bits into 4 registers of 8 bits. For scalable vectors, the number of output registers is unknown.

In the shufflevector example above, we know how many virtual registers we would like to unmerge into: One `(<vscale x 16 x s64>)` gets unmerged into two `(<vscale x 8 x s64>)`. We know this works because the minimum number of vector elements is correctly supported. In the original vector, the minimum is 16. In the two new vectors, the minimum is `8+8=16`. vscale is a compile time constant which will obviously result in the pre-unmerge vector having the same number of actual elements as the two unmerged vectors.

Now that there are two `(<vscale x 8 x s64>)` virtual registers, the same logic on how to map physical vectors onto scalable virtual registers follows the same process for any other instruction and is backend specific. This has nothing to do with G_MERGE/UNMERGE.

**Comment by tschuett** [comment]
Up to you, but to me it looks like misuse. I can unmerge  32bit into 4 8bit registers. But unmerging scalable vectors looks strange. Then you would need vector registers of different sizes. For AArch64 there are 32 scalable vector registers, Z0-Z31. They all have the same unknown scalable size.

**Comment by michaelmaitland** [comment]
> Up to you, but to me it looks like misuse. I can unmerge 32bit into 4 8bit registers. But unmerging scalable vectors looks strange. Then you would need vector registers of different sizes. For AArch64 there are 32 scalable vector registers, Z0-Z31. They all have the same unknown scalable size.

I am happy to have a discussion and see if we can come to an agreement.

I am not seeing a problem. As I see it, there is a contract that the destination types must be able to hold all the elements of the source type on an umerge (and the opposite for a merge). I believe that this is occurring with scalable vectors. I described this in my previous comment in more detail. The key is that vscale is a runtime constant which will be the same for both types, and the minimum number of elements is guaranteed to fit. Since the vscale is the same, the actual number of elements is guaranteed to fit.

> But unmerging scalable vectors looks strange. Then you would need vector registers of different sizes.

Isn't it the case that you would need vector registers of different sizes for fixed vectors too? For example, a `4 x i32` becomes two `2 x i32`. A `4 x i32` is a different size compared to `2 x i32`. For scalable vectors a `vscale x 4 x i32` becomes two `vscale x 2 x i32`. This is also two different sizes for the scalable vectors, since `vscale` is just a runtime constant.

Why is this a problem?

> For AArch64 there are 32 scalable vector registers, Z0-Z31. They all have the same unknown scalable size.

I'm not sure, but because you bring up physical registers, it sounds like you are having contention with the following
```
%0_:(vscale x N x sM)= G_IMPLICIT_DEF
$physicalVectorRegister = G_COPY %0 // How do I know that %0 fits in physicalVectorRegister
```
If this is the case, I don't think you have a problem with G_MERGE/UNMERGE. I think you have a problem with how we can map scalable vector typed virtual registers onto physical registers. On RISC-V, we have a mapping between LLVM types and physical registers. I don't think this is a discussion relevant to this PR, however I am happy to have it in a forum such as a GitHub issue.

CC: @topperc

File: llvm/lib/CodeGen/GlobalISel/Utils.cpp [hunk 1]
```diff
@@ -1071,49 +1071,73 @@ void llvm::getSelectionDAGFallbackAnalysisUsage(AnalysisUsage &AU) {
...
-    } else {
-      if (OrigElt.getSizeInBits() == TargetSize)
-        return OrigTy;
+    // TODO: The docstring for this function says the intention is to use this
+    // function to build MERGE/UNMERGE instructions. It won't be the case that
+    // we generate a MERGE/UNMERGE between fixed and scalable vector types. We
+    // could implement getLCMType between the two in the future if there was a
+    // need, but it is not worth it now as this function should not be used in
+    // that way.
+    if ((OrigTy.isScalableVector() && TargetTy.isFixedVector()) ||
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/Utils.cpp:1087]: Use an assert.

File: llvm/lib/CodeGen/GlobalISel/Utils.cpp [hunk 2]
```diff
@@ -1071,49 +1071,73 @@ void llvm::getSelectionDAGFallbackAnalysisUsage(AnalysisUsage &AU) {
...
+      // Prefer the original element type.
+      ElementCount Mul = OrigTy.getElementCount().multiplyCoefficientBy(
+          TargetTy.getElementCount().getKnownMinValue());
+      return LLT::vector(Mul.divideCoefficientBy(GCDMinElts),
+                         OrigTy.getElementType());
     }
-
-    unsigned LCMSize = std::lcm(OrigSize, TargetSize);
-    return LLT::fixed_vector(LCMSize / OrigElt.getSizeInBits(), OrigElt);
+    unsigned LCM = std::lcm(OrigTy.getElementCount().getKnownMinValue() *
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/Utils.cpp:1101]: Is `OrigTy.getElementCount().getKnownMinValue() * OrigElt.getSizeInBits().getFixedValue()` the same as `OrigTy.getSizeInBits().getKnownMinValue()`

Comment (michaelmaitland) [reply llvm/lib/CodeGen/GlobalISel/Utils.cpp:1101]: I don’t think so. I think SizeInBits is a TypeSize that represents ElementCount * ScalarSizeInBits

Comment (michaelmaitland) [reply llvm/lib/CodeGen/GlobalISel/Utils.cpp]: Yes, updated.

**Review by arsenm — APPROVED** [review]

**Comment by michaelmaitland** [comment]
> Up to you, but to me it looks like misuse.

I don't agree that it is a misuse so I am going to merge this. Happy to revisit the discussion down the line.

**Pull request merged**